cajita/src/Cajita_HypreSemiStructuredSolver.hpp

```diff
@@ -0,0 +1,1575 @@
+{
+};
+#endif // end HYPRE_USING_DEVICE_MEMORY
+
+//! Hypre device compatibility check - CUDA UVM memory.
+#ifdef HYPRE_USING_UNIFIED_MEMORY
+template <>
+struct HypreIsCompatibleWithMemorySpace<Kokkos::CudaUVMSpace> : std::true_type
+{
+};
+#endif // end HYPRE_USING_UNIFIED_MEMORY
+#endif // end KOKKOS_ENABLE_CUDA
+#endif // end HYPRE_USING_CUDA
+
+#ifdef HYPRE_USING_HIP
+#ifdef KOKKOS_ENABLE_HIP
+//! Hypre device compatibility check - HIP memory. FIXME - make this true when
+//! the HYPRE CMake includes HIP
+template <>
+struct HypreIsCompatibleWithMemorySpace<Kokkos::ExperimentalHIPSpace>
+    : std::false_type
+{
+};
+#endif // end KOKKOS_ENABLE_HIP
+#endif // end HYPRE_USING_HIP
+
+#ifndef HYPRE_USING_GPU
+//! Hypre device compatibility check - host memory.
+template <>
+struct HypreIsCompatibleWithMemorySpace<Kokkos::HostSpace> : std::true_type
+{
+};
+#endif // end HYPRE_USING_GPU
+
+*/
+
+//---------------------------------------------------------------------------//
+//! Hypre semi-structured solver interface for scalar fields.
+template <class Scalar, class EntityType, class MemorySpace>
+class HypreSemiStructuredSolver
+{
+  public:
+    //! Entity type.
+    using entity_type = EntityType;
+    //! Kokkos memory space..
+    using memory_space = MemorySpace;
+    //! Scalar value type.
+    using value_type = Scalar;
+    //! Object Type for SStruct
+    int object_type = HYPRE_SSTRUCT;
+    //! Hypre memory space compatibility check.
+    static_assert( HypreIsCompatibleWithMemorySpace<memory_space>::value,
+                   "HYPRE not compatible with solver memory space" );
+
+    /*!
+      \brief Constructor.
+      \param layout The array layout defining the vector space of the solver.
+      \param is_preconditioner Flag indicating if this solver will be used as
+      a preconditioner.
+    */
+    template <class ArrayLayout_t>
+    HypreSemiStructuredSolver( const ArrayLayout_t& layout, int n_vars,
+                           const bool is_preconditioner = false)
+        : _comm( layout.localGrid()->globalGrid().comm() )
+        , _is_preconditioner( is_preconditioner )
+    {
+        HYPRE_Init();
+
+        static_assert( is_array_layout<ArrayLayout_t>::value,
+                       "Must use an array layout" );
+        static_assert(
+            std::is_same<typename ArrayLayout_t::entity_type,
+                         entity_type>::value,
+            "Array layout entity type mush match solver entity type" );
+
+        // Spatial dimension.
+        const std::size_t num_space_dim = ArrayLayout_t::num_space_dim;
+
+        // Only a single part grid is supported initially
+        int n_parts = 1;
+        int part = 0;
+
+        // Only create data structures if this is not a preconditioner.
+        if ( !_is_preconditioner )
+        {
+            // Create the grid.
+            auto error = HYPRE_SStructGridCreate( _comm, num_space_dim, n_parts, &_grid );
```

Comment:
I just want to note that it would be great if hypre was also using template meta-programming and we could just reuse our previous interface and pass a new SStruct tag instead